build-system/eslint-rules/always-call-chai-methods.js

```diff
@@ -0,0 +1,170 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+// We need a full database of possible Chai assertions to figure out what to
+// lint for.
+const chai = require('chai');
+chai.use(require('chai-as-promised'));
+chai.use(require('sinon-chai'));
+// test/_init_tests.js
+chai.Assertion.addMethod('attribute');
+chai.Assertion.addMethod('class');
+chai.Assertion.addMethod('display');
+chai.Assertion.addMethod('jsonEqual');
+chai.Assertion.addProperty('visible');
+chai.Assertion.addProperty('hidden');
+
+const methods = [
+  ...Object.entries(Object.getOwnPropertyDescriptors(chai.Assertion.prototype)),
+]
+  // We're looking for method assertions, which are defined as real methods on the prototype.
+  // Property assertions (like `expect().to.be.ok`) are defined as getters.
+  .filter(([, desc]) => !!desc.value)
+  .map(([key]) => key);
+
+// Langauge chains don't actually assert anything, they're just there to make
+// sentences out of the tests.
+// https://github.com/chaijs/chai/blob/41ff363e26021433ae7e713b14c8f68fafc1c936/lib/chai/core/assertions.js#L13-L48
+const chaiLanguageChains = [
+  'to',
+  'be',
+  'been',
+  'is',
+  'that',
+  'which',
+  'and',
+  'has',
+  'have',
+  'with',
+  'at',
+  'of',
+  'same',
+  'but',
+  'does',
+  'still',
```

Comment:
Curious: Is it possible to import / infer this list from `node_modules/chai/lib/chai/core/assertions.js` so we don't have to manually maintain this list?

Reply:
For example, the following code...
```js
const chains = [
  ...Object.entries(Object.getOwnPropertyDescriptors(chai.Assertion.prototype)),
]
  .filter(
    ([, desc]) =>
      !desc.set && !!desc.get && desc.get.name == 'overwritingPropertyGetter'
  )
  .map(([key]) => key);
```
... gives us this list:
```js
chains [
  'to',        'be',         'been',
  'is',        'and',        'has',
  'have',      'with',       'that',
  'which',     'at',         'of',
  'same',      'but',        'does',
  'still',     'not',        'deep',
  'nested',    'own',        'ordered',
  'any',       'all',        'ok',
  'true',      'false',      'null',
  'undefined', 'NaN',        'exist',
  'empty',     'arguments',  'Arguments',
  'itself',    'extensible', 'sealed',
  'frozen',    'finite'
]
```

Perhaps there's a way to inspect `desc.get` and further filter down that list to the language chains that don't assert anything?

Reply:
Unfortunately not, I tried doing just this. There's no difference between a language chain and a property assertion externally.

Reply:
You could cheat and use the elements in `chains` up to but not including `'not'` so that newly added language chains will automatically be included. But I acknowledge that's ugly, and probably just as brittle as hard-coding the entire list.